File: src/domain/implementation-map/executable-stories.ts
/**
 * Executable stories filter.
 *
 * Determines which stories can be executed given the current
 * execution state, sorted by critical path priority.
 */
import type { ExecutionState, ParsedMap } from "./types.js";
import { StoryStatus } from "./types.js";

/** Check if a story is PENDING and has all dependencies satisfied (SUCCESS). */
function isExecutable(
  storyId: string,
  parsedMap: ParsedMap,
  state: ExecutionState,
): boolean {
  const storyState = state.stories[storyId];
  if (storyState === undefined) return false;
  if (storyState.status !== StoryStatus.PENDING) return false;

  const node = parsedMap.stories.get(storyId);
  if (node === undefined) return false;

  return node.blockedBy.every((depId) => {
    const depState = state.stories[depId];
    return depState !== undefined && depState.status === StoryStatus.SUCCESS;
  });
}

/** Sort stories: critical path first, then alphabetically. */
function sortByCriticalPath(
  stories: string[],
  parsedMap: ParsedMap,
): string[] {
  return stories.sort((a, b) => {
    const aOnCp = parsedMap.stories.get(a)?.isOnCriticalPath ?? false;
    const bOnCp = parsedMap.stories.get(b)?.isOnCriticalPath ?? false;
    if (aOnCp && !bOnCp) return -1;
    if (!aOnCp && bOnCp) return 1;
    return a.localeCompare(b);
  });
}

/** Get executable stories sorted by critical path priority. */
export function getExecutableStories(
  parsedMap: ParsedMap,
  executionState: ExecutionState,
): string[] {
  const executable: string[] = [];
  for (const storyId of parsedMap.stories.keys()) {
    if (isExecutable(storyId, parsedMap, executionState)) {
      executable.push(storyId);
    }
  }
  return sortByCriticalPath(executable, parsedMap);
}

File: java/src/main/java/dev/iadev/domain/implementationmap/ExecutableStories.java
package dev.iadev.domain.implementationmap;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Determines which stories are ready for execution given
 * the set of completed stories.
 *
 * <p>A story is executable when all its dependencies are
 * in the completed set. Results are sorted with critical
 * path stories first, then alphabetically.</p>
 */
public final class ExecutableStories {

    private ExecutableStories() {
    }

    /**
     * Filters stories that are ready for execution.
     *
     * @param dag              the DAG
     * @param completedStories set of story IDs already completed
     * @return sorted list of executable story IDs
     */
    public static List<String> filter(
            Map<String, DagNode> dag,
            Set<String> completedStories) {
        var executable = new ArrayList<String>();

        for (var entry : dag.entrySet()) {
            var id = entry.getKey();
            var node = entry.getValue();

            if (completedStories.contains(id)) {
                continue;
            }

            boolean allDepsComplete = node.blockedBy().stream()
                    .allMatch(completedStories::contains);

            if (allDepsComplete) {
                executable.add(id);
            }
        }

        executable.sort(criticalPathComparator(dag));
        return List.copyOf(executable);
    }

    private static Comparator<String> criticalPathComparator(
            Map<String, DagNode> dag) {
        return (a, b) -> {
            var nodeA = dag.get(a);
            var nodeB = dag.get(b);
            boolean aOnCp = nodeA != null
                    && nodeA.isOnCriticalPath();
            boolean bOnCp = nodeB != null
                    && nodeB.isOnCriticalPath();
            if (aOnCp && !bOnCp) {
                return -1;
            }
            if (!aOnCp && bOnCp) {
                return 1;
            }
            return a.compareTo(b);
        };
    }
}
